File: college/mydeliverables/phase1-ESA/for-submit/code/packages/progress/src/bayesian.ts
import type {
  SessionEvent,
  BayesianPosterior,
  RoleMultiplier,
  MaterialRole,
  TimeOfDay,
  ContextInsight,
} from './types'
import {
  BAYESIAN_PRIOR_MEAN,
  BAYESIAN_PRIOR_VARIANCE,
  MIN_SESSIONS_PER_BUCKET,
} from './config'
import { calibrationDenominator, isCalibrationSession } from './calibrationDenominator'

export interface HierarchicalResult {
  globalPosterior: BayesianPosterior
  globalMultiplier: number
  roleMultipliers: Partial<Record<MaterialRole, RoleMultiplier>>
  insights: ContextInsight[]
}

export function updatePosterior(
  prior: BayesianPosterior,
  observation: number,
  observationVariance: number,
): BayesianPosterior {
  const posteriorVariance =
    (prior.variance * observationVariance) /
    (prior.variance + observationVariance)
  const posteriorMean =
    (prior.variance * observation + observationVariance * prior.mean) /
    (prior.variance + observationVariance)
  return {
    mean: posteriorMean,
    variance: posteriorVariance,
    sessionCount: prior.sessionCount + 1,
  }
}

export function inferTimeOfDay(startedAt: string | undefined): TimeOfDay {
  if (!startedAt) return 'afternoon'
  const date = new Date(startedAt)
  const hour = date.getHours()
  if (hour < 12) return 'morning'
  if (hour < 17) return 'afternoon'
  return 'evening'
}

function computeEmpiricalVariance(ratios: number[]): number {
  if (ratios.length < 2) return BAYESIAN_PRIOR_VARIANCE
  const mean = ratios.reduce((s, r) => s + r, 0) / ratios.length
  const sumSq = ratios.reduce((s, r) => s + (r - mean) ** 2, 0)
  return Math.max(sumSq / (ratios.length - 1), 0.001)
}

export function computeHierarchicalModel(
  sessions: SessionEvent[],
  exceptionalIds: Set<string>,
): HierarchicalResult {
  const activeSessions = sessions.filter((s) => isCalibrationSession(s, exceptionalIds))

  if (activeSessions.length === 0) {
    return {
      globalPosterior: {
        mean: BAYESIAN_PRIOR_MEAN,
        variance: BAYESIAN_PRIOR_VARIANCE,
        sessionCount: 0,
      },
      globalMultiplier: BAYESIAN_PRIOR_MEAN,
      roleMultipliers: {},
      insights: [],
    }
  }

  const paceRatios = activeSessions.map(
    (s) => s.activeMinutes! / calibrationDenominator(s)!,
  )
  const observationVariance = computeEmpiricalVariance(paceRatios)

  // Level 0: Global posterior
  let globalPosterior: BayesianPosterior = {
    mean: BAYESIAN_PRIOR_MEAN,
    variance: BAYESIAN_PRIOR_VARIANCE,
    sessionCount: 0,
  }
  for (const ratio of paceRatios) {
    globalPosterior = updatePosterior(
      globalPosterior,
      ratio,
      observationVariance,
    )
  }

  // Level 1: Per-role posteriors (inherit global as prior)
  const roleGroups = new Map<MaterialRole, number[]>()
  for (const session of activeSessions) {
    if (session.materialRole) {
      const group = roleGroups.get(session.materialRole) ?? []
      group.push(session.activeMinutes! / calibrationDenominator(session)!)
      roleGroups.set(session.materialRole, group)
    }
  }

  const roleMultipliers: Partial<Record<MaterialRole, RoleMultiplier>> = {}
  for (const [role, ratios] of roleGroups) {
    if (ratios.length >= MIN_SESSIONS_PER_BUCKET) {
      const roleVariance = computeEmpiricalVariance(ratios)
      let rolePosterior: BayesianPosterior = {
        mean: globalPosterior.mean,
        variance: globalPosterior.variance,
        sessionCount: 0,
      }
      for (const ratio of ratios) {
        rolePosterior = updatePosterior(rolePosterior, ratio, roleVariance)
      }
      roleMultipliers[role] = {
        multiplier: rolePosterior.mean,
        confidence: 1 / Math.sqrt(rolePosterior.variance),
        sessionCount: ratios.length,
      }
    }
  }

  // Level 2: Per-role × time-of-day (insights only)
  const contextGroups = new Map<string, { role: MaterialRole; timeOfDay: TimeOfDay; ratios: number[] }>()
  for (const session of activeSessions) {
    if (session.materialRole) {
      const tod = inferTimeOfDay(session.startedAt)
      const key = `${session.materialRole}:${tod}`
      const group = contextGroups.get(key) ?? {
        role: session.materialRole,
        timeOfDay: tod,
        ratios: [],
      }
      group.ratios.push(session.activeMinutes! / calibrationDenominator(session)!)
      contextGroups.set(key, group)
    }
  }

  const insights: ContextInsight[] = []
  for (const [, group] of contextGroups) {
    if (group.ratios.length >= MIN_SESSIONS_PER_BUCKET) {
      const rolePrior = roleMultipliers[group.role]
      const priorMean = rolePrior ? rolePrior.multiplier : globalPosterior.mean
      const priorVariance = rolePrior
        ? 1 / (rolePrior.confidence ** 2)
        : globalPosterior.variance
      const contextVariance = computeEmpiricalVariance(group.ratios)

      let contextPosterior: BayesianPosterior = {
        mean: priorMean,
        variance: priorVariance,
        sessionCount: 0,
      }
      for (const ratio of group.ratios) {
        contextPosterior = updatePosterior(
          contextPosterior,
          ratio,
          contextVariance,
        )
      }

      const label = `${group.role} in the ${group.timeOfDay}`
      insights.push({
        role: group.role,
        timeOfDay: group.timeOfDay,
        multiplier: contextPosterior.mean,
        sessionCount: group.ratios.length,
        label,
      })
    }
  }

  return {
    globalPosterior,
    globalMultiplier: globalPosterior.mean,
    roleMultipliers,
    insights,
  }
}

File: college/mydeliverables/phase1-ESA/for-submit/code/packages/py-progress/src/py_progress/bayesian.py
from __future__ import annotations

from datetime import datetime

from py_progress.config import (
    BAYESIAN_PRIOR_MEAN,
    BAYESIAN_PRIOR_VARIANCE,
    MIN_SESSIONS_PER_BUCKET,
)
from py_progress.types import (
    BayesianPosterior,
    ContextInsight,
    DayOfWeek,
    HierarchicalResult,
    MaterialRole,
    RoleMultiplier,
    SessionEvent,
    TimeOfDay,
)


def update_posterior(
    prior: BayesianPosterior,
    observation: float,
    observation_variance: float,
) -> BayesianPosterior:
    posterior_variance = (prior.variance * observation_variance) / (
        prior.variance + observation_variance
    )
    posterior_mean = (
        prior.variance * observation + observation_variance * prior.mean
    ) / (prior.variance + observation_variance)
    return BayesianPosterior(
        mean=posterior_mean,
        variance=posterior_variance,
        sessionCount=prior.sessionCount + 1,
    )


def infer_time_of_day(started_at: str | None) -> TimeOfDay:
    if not started_at:
        return "afternoon"
    if started_at.endswith("Z"):
        date = datetime.fromisoformat(started_at.replace("Z", "+00:00"))
        hour = date.astimezone().hour
    else:
        hour = datetime.fromisoformat(started_at).hour
    if hour < 12:
        return "morning"
    if hour < 17:
        return "afternoon"
    return "evening"


def infer_day_of_week(started_at: str | None) -> DayOfWeek:
    if not started_at:
        return "weekday"
    if started_at.endswith("Z"):
        date = datetime.fromisoformat(started_at.replace("Z", "+00:00")).astimezone()
    else:
        date = datetime.fromisoformat(started_at)
    return "weekend" if date.weekday() >= 5 else "weekday"


def _compute_empirical_variance(ratios: list[float]) -> float:
    if len(ratios) < 2:
        return BAYESIAN_PRIOR_VARIANCE
    mean = sum(ratios) / len(ratios)
    sum_sq = sum((r - mean) ** 2 for r in ratios)
    return max(sum_sq / (len(ratios) - 1), 0.001)


def compute_hierarchical_model(
    sessions: list[SessionEvent],
    exceptional_ids: set[str],
) -> HierarchicalResult:
    active_sessions = [
        s
        for s in sessions
        if s.get("source") == "active"
        and s.get("plannedMinutes") is not None
        and s["plannedMinutes"] > 0
        and s.get("activeMinutes") is not None
        and s["activeMinutes"] > 0
        and (not s.get("sessionId") or s["sessionId"] not in exceptional_ids)
    ]

    if not active_sessions:
        return HierarchicalResult(
            globalPosterior=BayesianPosterior(
                mean=BAYESIAN_PRIOR_MEAN,
                variance=BAYESIAN_PRIOR_VARIANCE,
                sessionCount=0,
            ),
            globalMultiplier=BAYESIAN_PRIOR_MEAN,
            roleMultipliers={},
            insights=[],
        )

    pace_ratios = [s["activeMinutes"] / s["plannedMinutes"] for s in active_sessions]
    observation_variance = _compute_empirical_variance(pace_ratios)

    global_posterior = BayesianPosterior(
        mean=BAYESIAN_PRIOR_MEAN,
        variance=BAYESIAN_PRIOR_VARIANCE,
        sessionCount=0,
    )
    for ratio in pace_ratios:
        global_posterior = update_posterior(
            global_posterior, ratio, observation_variance
        )

    role_groups: dict[MaterialRole, list[float]] = {}
    for session in active_sessions:
        role = session.get("materialRole")
        if role:
            role_groups.setdefault(role, []).append(
                session["activeMinutes"] / session["plannedMinutes"]
            )

    role_multipliers: dict[str, RoleMultiplier] = {}
    for role, ratios in role_groups.items():
        if len(ratios) >= MIN_SESSIONS_PER_BUCKET:
            role_variance = _compute_empirical_variance(ratios)
            role_posterior = BayesianPosterior(
                mean=global_posterior.mean,
                variance=global_posterior.variance,
                sessionCount=0,
            )
            for ratio in ratios:
                role_posterior = update_posterior(role_posterior, ratio, role_variance)
            role_multipliers[role] = RoleMultiplier(
                multiplier=role_posterior.mean,
                confidence=1 / (role_posterior.variance**0.5),
                sessionCount=len(ratios),
            )

    context_groups: dict[str, dict] = {}
    for session in active_sessions:
        role = session.get("materialRole")
        if role:
            tod = infer_time_of_day(session.get("startedAt"))
            key = f"{role}:{tod}"
            group = context_groups.setdefault(
                key, {"role": role, "timeOfDay": tod, "ratios": []}
            )
            group["ratios"].append(
                session["activeMinutes"] / session["plannedMinutes"]
            )

    insights: list[ContextInsight] = []
    for group in context_groups.values():
        ratios = group["ratios"]
        if len(ratios) >= MIN_SESSIONS_PER_BUCKET:
            role_prior = role_multipliers.get(group["role"])
            prior_mean = role_prior.multiplier if role_prior else global_posterior.mean
            prior_variance = (
                1 / (role_prior.confidence**2)
                if role_prior
                else global_posterior.variance
            )
            context_variance = _compute_empirical_variance(ratios)

            context_posterior = BayesianPosterior(
                mean=prior_mean,
                variance=prior_variance,
                sessionCount=0,
            )
            for ratio in ratios:
                context_posterior = update_posterior(
                    context_posterior, ratio, context_variance
                )

            label = f"{group['role']} in the {group['timeOfDay']}"
            insights.append(
                ContextInsight(
                    role=group["role"],
                    timeOfDay=group["timeOfDay"],
                    multiplier=context_posterior.mean,
                    sessionCount=len(ratios),
                    label=label,
                )
            )

    return HierarchicalResult(
        globalPosterior=global_posterior,
        globalMultiplier=global_posterior.mean,
        roleMultipliers=role_multipliers,
        insights=insights,
    )
